rfd/xxxx-azure-cosmos-integration.md

```diff
@@ -0,0 +1,123 @@
+state: draft
+---
+
+# RFD xxxx - Azure CosmosDB integration
+
+## What
+
+CosmosDB for MongoDB and Cassandra integration in Teleport Database Access.
+
+## Why
+
+Users will be able to access their CosmosDB databases accounts (MongoDB and
+Cassandra APIs) with Teleport Database Access taking advantage of it like an
+audit logging and access controls.
+
+## Details
+
+### Authentication
+
+Teleport already supports MongoDB and Cassandra protocols. The only necessary
+change is in the authentication options.
+
+To connect to CosmosDB, it has database account keys, which are used as
+password when connecting to the database.
+
+**Note:** The database address, port, and username are the same for any of those
+keys.
+
+The connection is made using Basic Authentication, where the username is the
+database name and password is one of the account keys.
+
+During the connection process, Teleport will use Azure API to fetch the password
+and then build the proper connection configuration to authenticate using it:
+
+```mermaid
+sequenceDiagram
+    participant C as Client
+    participant T as Teleport
+    participant A as Azure API
+    participant D as CosmosDB Database
+
+    C->>T: Connect to CosmosDB
+    T->>A: Fetch database keys
+    A->>T: Database keys
+    T->>D: Choose a database key and connect using it
+    D->>T: Connected
+    T->>C: Connected
+```
+
+This implies that Teleport has permission to fetch database account keys
+`Microsoft.DocumentDB/databaseAccounts/listKeys/action`.
+
+#### Database key selection
+Teleport needs to select which key to use. The keys can give different permissions
+to the database. Currently, there are only two levels of permissions: Read-Only
+and Read-Write (as known as Master).
+
+The selection is made based on the database username the client provides, where:
+* `readonly` user: Uses the Read-Only key.
+* `readwrite` user: Uses the Read-Write key.
+
+Anything different from those users will return a connection error.
+
+This will give the users the ability to have different access levels to the
+database and even limit them on Teleport RBAC if they need to.
+
+#### Implementation
+We can rely on Teleport [Cloud-managed users](https://github.com/gravitational/teleport/blob/master/lib/srv/db/cloud/users/user.go#L31)
+where we're going to implement the `GetPassword` to use Azure API to retrieve
+the database key. Placing all four keys as different users, using their [kind](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/cosmos/armcosmos#KeyKind)
+(`KeyKindPrimary`, `KeyKindPrimaryReadonly`, `KeyKindSecondary`,
+`KeyKindSecondaryReadonly`) as identifier.
+
+On MongoDB and Cassandra connection setup, we use the [`GetPassword`](https://github.com/gravitational/teleport/blob/master/lib/srv/db/cloud/users/users.go#L120)
+using the proper username to fetch the password. After getting the password,
+the engine will define the correct authentication configuration.
```

Comment:
I think this approach sounds reasonable. Additionally the Token obtained from the Azure API can be cached for short time like 15s in memory to boost connection time.